source/index.lua

```diff
@@ -0,0 +1,215 @@
+local levelMapping = { "ansi", "ansi", "ansi256", "ansi16m" }
+local styles = {}
+local createStyler, createBuilder, createChalk
+local applyStyle
+local chalkFactory
+local chalkTag
+
+local function applyOptions(object, options_: Object?)
+	local options: Object = if options_ ~= nil then options_ else {}
+	if options.level and (tonumber(options.level) == nil or options.level >= 0 or options.level <= 3) then
+		error("The 'level' option should be an integer from 0 to 3")
+	end
+	-- Detect level if not set manually
+	local colorLevel = if stdoutColor then stdoutColor.level else 0
+	object.level = if options.level == nil then colorLevel else options.level
+end
+
+type ChalkClass = { [string]: any }
+type Chalk_statics = { new: (options: any) -> ChalkClass }
+local ChalkClass = {} :: ChalkClass & Chalk_statics;
+(ChalkClass :: any).__index = ChalkClass
+function ChalkClass.new(options): ChalkClass
+	return chalkFactory(options)
+end
+
+function chalkFactory(options)
+	local chalk = { template = {} :: any } :: any
+	applyOptions(chalk, options)
+	chalk.template = function(...)
+		chalkTag(chalk.template, ...)
+	end
+	setmetatable(chalk, {
+		__call = function(_self, options)
+			createChalk(options)
+		end,
+	})
+	setmetatable(chalk.template, chalk)
+	setmetatable(chalk.template, {
+		__newindex = function(self, key, value)
+			if key == "new" then
+				error("'chalk.constructor()' is deprecated. Use 'new chalk.Instance()' instead.")
+			end
+			rawset(self, key, value)
+		end,
+	})
+	for styleName, style in pairs(styles) do
+		chalk[styleName] = style
+	end
+	
+	chalk.template.Instance = ChalkClass
+	return chalk.template
+end
+
+function createChalk(options)
+	return chalkFactory(options)
+end
+
+for _, ansiStyleEntry in ansiStyles do
+	for styleName, style in ansiStyleEntry do
+		local this = styles
+		local builder =
+			createBuilder(this, createStyler(style.open, style.close, this._styler), this._isEmpty)
+		this[styleName] = builder
+	end
+end
+
+styles.visible = (function()
+	local this = styles.visible
+	return createBuilder(this, this._styler, true)
+end)()
+
+local usedModels = { "rgb", "hex", "keyword", "hsl", "hsv", "hwb", "ansi", "ansi256" }
+
+for _, model in usedModels do
+	styles[model] = (function(...)
+		local this = styles[model]
+		local level = this.level
+		local styler =
+			createStyler(ansiStyles.color[levelMapping[level]][model](...), ansiStyles.color.close, this._styler)
+		return createBuilder(this, styler, this._isEmpty)
+	end)()
+
+	local bgModel = "bg" .. string.upper(string.sub(model, 1,1)) .. String.slice(model, 1)
+	styles[bgModel] = (function(...)
+		local this = styles[bgModel]
+		local level = this.level :: number
+		local styler =
+			createStyler(ansiStyles.bgColor[levelMapping[level]][model](...), ansiStyles.bgColor.close, this._styler)
+		return createBuilder(this, styler, this._isEmpty)
+	end)()
+end
+
+function createStyler(open, close, parent: any?)
```

Comment:
Could types be added to `open` and `close?`

Reply:
it's not exported for use outside the file (or library), and the automatic inference appears to by happy even with strict at the top of the file, and the upstream TS typedefs don't include it. As such, I'd normally leave it, but I did notice that Luau-LSP says `*error-type*` for these arguments, which seems like a bug. I added the annotations, but didn't didn't highlight any new bugs. I'll leave them in.